core/commonMain/src/kotlinx/serialization/builtins/InstantComponentSerializer.kt

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2025-2025 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.serialization.builtins
+
+import kotlinx.serialization.*
+import kotlinx.serialization.descriptors.*
+import kotlinx.serialization.encoding.*
+import kotlin.time.ExperimentalTime
+import kotlin.time.Instant
+
+/**
+ * Serializer that encodes and decodes [Instant] as its second and nanosecond components of the Unix time.
+ *
+ * JSON example: `{"epochSeconds":1607505416,"nanosecondsOfSecond":124000}`.
+ */
+@ExperimentalTime
+public object InstantComponentSerializer : KSerializer<Instant> {
+
+    override val descriptor: SerialDescriptor =
+        buildClassSerialDescriptor("kotlinx.serialization.InstantComponentSerializer") {
+            element<Long>("epochSeconds")
+            element<Long>("nanosecondsOfSecond", isOptional = true)
+        }
+
+    @OptIn(ExperimentalSerializationApi::class)
+    override fun deserialize(decoder: Decoder): Instant =
+        decoder.decodeStructure(descriptor) {
+            var epochSeconds: Long? = null
+            var nanosecondsOfSecond = 0
+            while (true) {
+                when (val index = decodeElementIndex(descriptor)) {
+                    0 -> epochSeconds = decodeLongElement(descriptor, 0)
+                    1 -> nanosecondsOfSecond = decodeIntElement(descriptor, 1)
+                    CompositeDecoder.DECODE_DONE -> break
+                    else -> throw SerializationException("Unexpected index: $index")
+                }
+            }
+            if (epochSeconds == null) throw MissingFieldException(
```

Comment:
To avoid boxing it may be better to either have a separate primitive value that records whether epochSeconds was seen, or have a "sentinel" value that is not a valid Instant (e.g. `Long.MAX` or `Long.MIN`)